### PR TITLE
[ErrorHandler] Add support for `@return-type-will-change`

### LINKED ISSUE #58128: [ErrorHandler] DebugClassloader: introduce an annotation to signal definite intent to add/change the return typehint of a method
### Description

Right now, DebugClassloader triggers a deprecation when a method with no return typehint implements/extends from a method that has in the docblock a `@return` annotation that specifies a type.

This has 2 limitations at the moment:

1. the check is limited to extending/implementing classes in a different vendor only
2. the deprecation is only 'potential' i.e. there is no certainty if/when it will happen, as it may be related to BC considerations etc

In Drupal we have a large legacy of unhinted code, and these limitations make difficult for us to engage in a managed journey to introduce return typehints. See https://drupal.slack.com/archives/C1BMUQ9U6/p1724921060882849 for some discussion.

We would suggest to introduce an annotation, say `@return-type-will-change`, to signal to DebugClassloader that the change WILL positively happen, so that deprecation triggered can be more effective in pressing for typehinting of the extending/implementing classes before next release; that deprecation can then be triggered also for classes in the same vendor.

### Example

For example, in an interface file have this: 

```php
/**
 * Returns the height of the image.
 *
 * @return int|null
 *   The height of the image, or NULL if the image is invalid.
 *
 * @return-type-will-change in drupal:12.0.0
 *
 * @todo Uncomment return type hint in drupal:12.0.0.
 * @see https://www.drupal.org/project/drupal/issues/3470951
 */
public function getHeight()/* : ?int */;
```

and then any class whose method is not typehinted will trigger a deprecation like

`Method "%s::%s()" will add "%s" as a native return type declaration in version 12.0.0. Implement the typehint in %s "%s" now to avoid errors.`

### PR DESCRIPTION
**Pull request by mondrake**
| Q             | A
| ------------- | ---
| Branch?       | 7.2
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Issues        | Fix #58128
| License       | MIT

See
* #58128

### CODE REVIEW

**Review by pan93412 — APPROVED** [review]

File: src/Symfony/Component/ErrorHandler/DebugClassLoader.php [hunk 1]
```diff
@@ -606,6 +615,10 @@ public function checkAnnotations(\ReflectionClass $refl, string $class): array
                 }
             }
 
+            if (isset($doc['return-typehint-will-change'])) {
```
Comment (stof) [inline src/Symfony/Component/ErrorHandler/DebugClassLoader.php]: this should say `type`, not `typehint`, to match the way it is named in PHP (PHP does not have a concept of typehint, but of type)

Comment (mondrake) [reply src/Symfony/Component/ErrorHandler/DebugClassLoader.php]: fixed

**Comment by mondrake** [comment]
Thanks for reviewing @stof. Did that, and changed the way the string is built if any text is passed with the annotation. Thought projects may have different requirements as to the string and better leave to their code sniffers to validate it if needed.

**Comment by nicolas-grekas** [comment]
Sorry I still don't get it. From the linked issue:

> having both  #[ReturnTypeWillChange] and @return together: if I understand well it will be no good for Drupal's BC - that would mean actually doing the change on the interface, then adding  #[ReturnTypeWillChange] to the implementing methods. 

What prevents adding `#[ReturnTypeWillChange]` and `@return` together on *interfaces*? Wouldn't that be the explicit opt-in you're looking for? And doesn't this work already?

**Comment by stof** [comment]
@nicolas-grekas the meaning of `ReturnTypeWillChange` in PHP is that we acknowledge that we will change our own return type later, before migrating to a version when the parent forces us to have one. It is an attribute meant to opt-out of a deprecation warning triggered on the method where it is applied.

Repurposing the same attribute to indicate that the phpdoc return type should be considered a tentative return type might cause some confusion in the ecosystem.

> And doesn't this work already?

AFAICT, the issue is that DebugClassLoader considers that any `@return` is indicating an intent of migrating to a native return type (and so needing a deprecation for child class). There is currently no opt-in for which phpdoc return types are tentative return types.

**Comment by mondrake** [comment]
If it helps, we might change the annotation to be `@return-type-will-be-added` which is in fact what this PR is doing, and avoid confusion with the PHP native attribute.

**Comment by nicolas-grekas** [comment]
>the meaning of ReturnTypeWillChange in PHP is that we acknowledge that we will change our own return type later, before migrating to a version when the parent forces us to have one. It is an attribute meant to opt-out of a deprecation warning triggered on the method where it is applied.

That's maybe where our interpretation differs.
To me, `@return-type-will-be-added` can be expressed by adding both an `@return` annotation and a `#[ReturnTypeWillChange]` attribute.

I think this is safe assumption, because to me the difference you describe @stof doesn't exist: both interpretations are essentially equivalent.

**Comment by mondrake** [comment]
@nicolas-grekas @stof @catch56 just a couple of points since this is originally coming from findings in Drupal.

Regardless of the interpretations, the current state falls short for Drupal for two reasons:
- Drupal needs a way to communicate changing/added return type for its own vendor namespace - that is not working: only classes from a vendor namespace different than the one of the declaring class are currently returning the deprecation.
- Drupal is usually POSITIVE and not POTENTIAL on deprecations, and indicates the specific version it will no longer support a feature/method/you-name-it.  The difference is 'WILL' vs 'MIGHT'.

**Comment by nicolas-grekas** [comment]
Would it be possible to build this on the ReturnTypeWillChange attribute?
I had a quick look at the code and we don't do anything at the moment, but it would make sense to me to consider a method with a ReturnTypeWillChange as having a tentative return type (declared with the `@return` annotation).

**Comment by mondrake** [comment]
In Drupal we would not need that. If I understand correctly, we would have to find all the possible instances that will change BEFORE we do the change. What we need to do is to signal top down the change will happen, and let any extenders react on the upcoming change. With the module concept in Drupal, we cannot control all modules beforehand.

**Comment by catch56** [comment]
Yes the workflow we're trying to implement is:

1. Drupal core decides it's going to add return type hints somewhere
2. Drupal core indicates to contributed modules (which are in the \Drupal namespace but with separate maintainers) that a specific method will have a return type hint added in the next major release.
3. Contrib maintainers (hopefully) update their interface/abstract class etc. implementations to add the return type hint.
4. Drupal core actually adds the real return type hint in the next major release.

I understand ReturnTypeWillChange to be for the implementing class rather than the parent class/interface usually.

**Comment by mondrake** [comment]
Given that this seems controversial, I started working on a different approach, based on a PHPStan rule. That way we would also be able to baseline the errors.

**Comment by nicolas-grekas** [comment]
What's controversial is the annotation. I pushed a commit where the existing attribute is used instead.

> I understand ReturnTypeWillChange to be for the implementing class rather than the parent class/interface usually.

Sorry I don't get why. To me this works perfectly fine on parent classes.

**Comment by mondrake** [comment]
From PHP documentation, https://www.php.net/manual/en/class.returntypewillchange.php:

> [some] methods now **require overriding methods to declare a compatible return type**, otherwise a deprecated notice is emitted during inheritance validation. In case the return type cannot be declared for an overriding method due to PHP cross-version compatibility concerns, a #[\ReturnTypeWillChange] attribute can be added to **silence the deprecation notice**.

(emphasis mine)

This tells me 3 things:

1. the attribute is to be used in **overriding methods**
2. it is used to **silence the deprecation**
3. it is used for PHP cross-version compatibility, not userland

This is where the controversy/confusion sparks, I think: the usage of the attribute in this last commit implies that it is used in the _interface_ (parent method), and it is used to _trigger_  the deprecation. This is 180° change of directions to what people commonly understand about the attribute.

My 2 cents...

**Comment by catch56** [comment]
Yes I like the idea of an attribute for this but I'm not sure about re-using the PHP native one, if they ever change the scope of it it could conflict.

**Comment by nicolas-grekas** [comment]
What  understand about the attribute: it means what it names says, aka "this method is going to have a native return type in the future". And because the author of the method said so, PHP will stop annoying them when overriding an internal method that also has a tentative return type, because well, the author promised it will add one soon. This side-effect for the php-engine is just a corollary of what the attribute declares: this method is going to have a return type soon.

**Comment by mondrake** [comment]
The PHPStan custom rule approach is outlined in the [Introduce a @return-type-will-be-added annotation to indicate return type addition in next major; enforce with a PHPStan rule](https://www.drupal.org/project/drupal/issues/3486376) Drupal.org issue and its MR.

**Comment by nicolas-grekas** [comment]
OK, let me close.
I think this approach with the attribute would be useful because it works in more situations.
In addition to the phpstan rule, that'd give better visibility to the deprecations.
But looking at how the discussion went, you won't use it, and we don't need to merge things that people don't use so let's close.

**Comment by mondrake** [comment]
Thank you for the reviews and feedback!

**Pull request closed without merging**